ts/src/core/clause.ts

```diff
@@ -1371,6 +1377,65 @@ export function PaginationMultipleColsSubQuery<T extends Data, K = keyof T>(
   );
 }
 
+export type PaginationUnboundColsQueryOrdering<T extends Data, K = keyof T> = {
+  sortCol: K;
+  direction: "ASC" | "DESC";
+  sortValue: string | number | null;
+  nullsPlacement?: "first" | "last";
+  overrideAlias?: string;
+};
+
+/**
+ * When paginating over multiple ordered columns, we need to construct a nested
+ * set of clauses to correctly filter. The resulting query will be structured as
+ * a group of nested ANDs and ORs.
+ *
+ * For example, given a set of ordering [{col1, less1}, {col2, greater2}], the
+ * function will construct a query that looks like:
+ *
+ * (col1 < $1 OR (col1 = $1 AND col2 > $2))
+ */
+export function PaginationUnboundColsQuery<T extends Data, K = keyof T>(
+  ordering: PaginationUnboundColsQueryOrdering<T, K>[],
+): Clause<T, K> | undefined {
+  if (ordering.length === 0) {
+    throw new Error("Must provide at least one ordering.");
+  }
+  let nesting: Clause<T, K> | undefined;
+  ordering
+    .reverse()
+    .forEach(
+      ({ sortCol, direction, sortValue, nullsPlacement, overrideAlias }) => {
+        const nullsOrder =
+          nullsPlacement ?? (direction === "DESC" ? "first" : "last");
+        const clauseFn = direction === "DESC" ? Less : Greater;
+        const baseClause = clauseFn(sortCol, sortValue, overrideAlias);
+        const withNullsClause =
+          sortValue !== null
+            ? nullsOrder === "last"
+              ? Or(baseClause, Eq(sortCol, null, overrideAlias))
+              : And(baseClause, NotEq(sortCol, null, overrideAlias))
+            : nullsOrder === "last"
+              ? Eq(sortCol, null, overrideAlias)
+              : undefined; // If nulls first and value is null, can't filter here
```

Comment:
still feels like we need more comments here

Reply:
Yeah I can do that